In [4]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
#MORE LIBRARIES DEFINED WHEN WE DID TEXT PRE PROCESSING


In [5]:
df_1 = pd.read_csv('reviews.csv', sep='\t')

In [6]:
df_1.head()

Name  RatingValue DatePublished  \
0  Playa Cabana            4    2020-02-26   
1  Playa Cabana            3    2019-08-04   
2  Playa Cabana            3    2019-08-24   
3  Playa Cabana            4    2019-06-06   
4  Playa Cabana            5    2020-05-25   

                                              Review  
0  I was tasked with finding a spot for a group d...  
1  Went here with my friends and family. I liked ...  
2  Surprisingly good Flautas! They came as 3 roll...  
3  As a Mexican I always crave authentic Mexican ...  
4  Best tacos I've ever had. Both locations are g...

In [7]:
df_2 = df_1.copy()

In [8]:
#dictionary inside the dictionary
df_2 = df_2.replace({'RatingValue' : { 5:"positive",
                                      4:"positive",
                                      3:"neutral",
                                      2:"negative",
                                      1:"negative"}})
df_2.head()
                                      
                                     

Name RatingValue DatePublished  \
0  Playa Cabana    positive    2020-02-26   
1  Playa Cabana     neutral    2019-08-04   
2  Playa Cabana     neutral    2019-08-24   
3  Playa Cabana    positive    2019-06-06   
4  Playa Cabana    positive    2020-05-25   

                                              Review  
0  I was tasked with finding a spot for a group d...  
1  Went here with my friends and family. I liked ...  
2  Surprisingly good Flautas! They came as 3 roll...  
3  As a Mexican I always crave authentic Mexican ...  
4  Best tacos I've ever had. Both locations are g...

In [9]:
df_2.shape

(1920, 4)

In [10]:
#data description
#print(df_2.describe())

In [11]:
#column name
print(df_2.columns)

Index(['Name', 'RatingValue', 'DatePublished', 'Review'], dtype='object')


In [12]:
#data type
print(df_2.dtypes)

Name             object
RatingValue      object
DatePublished    object
Review           object
dtype: object


In [13]:
#the summary of the categorical values

categorical = df_2.dtypes[df_2.dtypes == "object"].index
print(categorical)
df_2[categorical].describe()

Index(['Name', 'RatingValue', 'DatePublished', 'Review'], dtype='object')


Name RatingValue DatePublished  \
count              1920        1920          1920   
unique               96           3           767   
top     Nodo Restaurant    positive    2020-02-16   
freq                 40        1465            13   

                                                   Review  
count                                                1920  
unique                                               1900  
top     These tacosssss. Hello. Still dreaming about t...  
freq                                                    2

In [14]:
#CHALLANGES-CHECK MISSING VALUES

df_2.isna().sum()

Name             0
RatingValue      0
DatePublished    0
Review           0
dtype: int64

In [15]:
List_Of_Categories_In_Column=list(df_2['RatingValue'].value_counts().index)
List_Of_Categories_In_Column

['positive', 'neutral', 'negative']

In [16]:
List_Of_Categories_In_Column_with_counts=list(df_2['RatingValue'].value_counts())
List_Of_Categories_In_Column_with_counts

[1465, 297, 158]

In [17]:
df_positive = df_2[df_2["RatingValue"] =="positive"]
df_positive = df_positive.copy()
df_positive_new = df_positive.head(158)
print(df_positive_new.shape)
print(df_positive_new.head())


(158, 4)
           Name RatingValue DatePublished  \
0  Playa Cabana    positive    2020-02-26   
3  Playa Cabana    positive    2019-06-06   
4  Playa Cabana    positive    2020-05-25   
5  Playa Cabana    positive    2019-08-31   
6  Playa Cabana    positive    2019-04-07   

                                              Review  
0  I was tasked with finding a spot for a group d...  
3  As a Mexican I always crave authentic Mexican ...  
4  Best tacos I've ever had. Both locations are g...  
5  Food was exceptional and service was pretty fa...  
6  Their space is a bit cramped, but Playa Cabana...  


In [18]:
df_neutral = df_2[df_2["RatingValue"] == "neutral"]
df_neutral = df_neutral.copy()
df_neutral_new = df_neutral.head(158)
print(df_neutral_new.shape)
print(df_neutral_new.head())

(158, 4)
               Name RatingValue DatePublished  \
1      Playa Cabana     neutral    2019-08-04   
2      Playa Cabana     neutral    2019-08-24   
12     Playa Cabana     neutral    2018-09-07   
16     Playa Cabana     neutral    2018-08-13   
30  Wilbur Mexicana     neutral    2019-12-14   

                                               Review  
1   Went here with my friends and family. I liked ...  
2   Surprisingly good Flautas! They came as 3 roll...  
12  My friend and I decided on Mexican food for di...  
16  I've been duped by yelp reviews!! We were give...  
30  Great spot if you wanna get some quick tacos o...  


In [19]:
df_negative = df_2[df_2["RatingValue"] == "negative"]
df_negative =df_negative.copy()
df_negative_new = df_negative
print(df_negative_new.shape)
print(df_negative_new.head())

(158, 4)
            Name RatingValue DatePublished  \
53  Como En Casa    negative    2019-02-08   
57  Como En Casa    negative    2019-08-27   
58  Como En Casa    negative    2018-03-27   
59  Como En Casa    negative    2019-07-05   
70  Mexico Lindo    negative    2018-06-06   

                                               Review  
53  $16 dollars for the worst burrito bowl/platter...  
57  I took the advice of one of the reviewers on Y...  
58  Two words can summarize my thoughts on the foo...  
59  Based on the ratings and reviews I was very ex...  
70  This place has been bookmarked for awhile and ...  


In [20]:
#now concatenate all three dataframes
data_frame = [df_negative_new ,df_positive_new,df_neutral_new]
result = pd.concat(data_frame)
result



Name RatingValue DatePublished  \
53             Como En Casa    negative    2019-02-08   
57             Como En Casa    negative    2019-08-27   
58             Como En Casa    negative    2018-03-27   
59             Como En Casa    negative    2019-07-05   
70             Mexico Lindo    negative    2018-06-06   
...                     ...         ...           ...   
1048                Levetto     neutral    2018-11-17   
1054  La Vecchia Ristorante     neutral    2019-08-15   
1057  La Vecchia Ristorante     neutral    2019-07-18   
1071  La Vecchia Ristorante     neutral    2018-08-29   
1075       Taverna Mercatto     neutral    2019-10-24   

                                                 Review  
53    $16 dollars for the worst burrito bowl/platter...  
57    I took the advice of one of the reviewers on Y...  
58    Two words can summarize my thoughts on the foo...  
59    Based on the ratings and reviews I was very ex...  
70    This place has been bookmarked for awhile and ...  
...                                                 ...  
1048  Came on a Saturday evening at 5:30 and it was ...  
1054  La Vecchia is a cute restaurant with great amb...  
1057  Going to start off this review by saying the f...  
1071  My family and I decided to choose La Vecchia f...  
1075  A last minute decision as we were staying at D...  

[474 rows x 4 columns]

In [21]:
from sklearn.utils import shuffle

In [22]:
df = shuffle(result)


In [23]:
df.reset_index(inplace=True, drop=True)

# 1)Dataset

In [24]:
df.head()

Name RatingValue DatePublished  \
0     Playa Cabana     neutral    2018-09-07   
1       Fieramosca     neutral    2018-06-19   
2  El Nahual Tacos    negative    2019-09-10   
3             Miku    negative    2019-12-10   
4     Good Hombres    positive    2019-11-10   

                                              Review  
0  My friend and I decided on Mexican food for di...  
1  Had a good dinner here last week! The interior...  
2  To be completely honest. I was expecting more....  
3  I wanted sooo badly to love Miku ...Because of...  
4  This place serves some really great tacos! Ser...

In [25]:
#to further check the dataset is balanced or not

#Lets see if Length of reviews has any association with Label Classification
df['length']=df['Review'].apply(len) #adding a length Column

C:\Users\Mahwish\Anaconda2020\lib\site-packages\ipykernel_launcher.py:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  after removing the cwd from sys.path.


In [26]:
df.head() #viewing top 5 rows

Name RatingValue DatePublished  \
0     Playa Cabana     neutral    2018-09-07   
1       Fieramosca     neutral    2018-06-19   
2  El Nahual Tacos    negative    2019-09-10   
3             Miku    negative    2019-12-10   
4     Good Hombres    positive    2019-11-10   

                                              Review  length  
0  My friend and I decided on Mexican food for di...    1153  
1  Had a good dinner here last week! The interior...    1939  
2  To be completely honest. I was expecting more....     393  
3  I wanted sooo badly to love Miku ...Because of...    1370  
4  This place serves some really great tacos! Ser...     241

In [27]:
#Summary Statistics
#Lets generate Summary Statistics of Review Length group by classes
#Also to further check the dataset is balanced or not

#df[['length','RatingValue']].groupby('RatingValue').describe()

# 2)Analysing the reviews

### Word Count – total number of words in each review

In [28]:
# Calculate word count
df['word_count'] = df['Review'].apply(lambda x: len(str(x).split(" ")))


C:\Users\Mahwish\Anaconda2020\lib\site-packages\ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  


### Character Count – total number of characters in each review

In [29]:
# Calculate character count
df['char_count'] = df['Review'].str.len()

C:\Users\Mahwish\Anaconda2020\lib\site-packages\ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  


### Average word length – the average length of words used

In [30]:
def avg_word(Review):
  words = Review.split()
  return (sum(len(word) for word in words) / len(words))

# Calculate average words
df['avg_word'] = df['Review'].apply(lambda x: avg_word(x))

C:\Users\Mahwish\Anaconda2020\lib\site-packages\ipykernel_launcher.py:6: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  


### Stopword Count – total number of words which are considered stop words

In [31]:
# Import stopwords
from nltk.corpus import stopwords

### Calculate number of stop words

In [32]:
stop_words = stopwords.words('english')
df['stopword_coun'] = df['Review'].apply(lambda x: len([x for x in x.split() if x in stop_words]))

C:\Users\Mahwish\Anaconda2020\lib\site-packages\ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  


In [33]:
df.describe()

length  word_count   char_count    avg_word  stopword_coun
count   474.000000  474.000000   474.000000  474.000000     474.000000
mean    799.419831  146.158228   799.419831    4.442453      59.529536
std     654.140443  119.416863   654.140443    0.316498      49.947924
min      87.000000   13.000000    87.000000    3.631579       3.000000
25%     368.250000   67.000000   368.250000    4.244850      27.000000
50%     610.000000  112.000000   610.000000    4.420964      45.500000
75%    1057.500000  198.750000  1057.500000    4.620898      81.750000
max    5000.000000  944.000000  5000.000000    5.857143     366.000000

In [34]:
df.head()

Name RatingValue DatePublished  \
0     Playa Cabana     neutral    2018-09-07   
1       Fieramosca     neutral    2018-06-19   
2  El Nahual Tacos    negative    2019-09-10   
3             Miku    negative    2019-12-10   
4     Good Hombres    positive    2019-11-10   

                                              Review  length  word_count  \
0  My friend and I decided on Mexican food for di...    1153         209   
1  Had a good dinner here last week! The interior...    1939         370   
2  To be completely honest. I was expecting more....     393          69   
3  I wanted sooo badly to love Miku ...Because of...    1370         255   
4  This place serves some really great tacos! Ser...     241          45   

   char_count  avg_word  stopword_coun  
0        1153  4.410377             92  
1        1939  4.204301            175  
2         393  4.628571             24  
3        1370  4.319066            118  
4         241  4.377778             21

# 3)Cleaning  the Data Set

How does the old saying go?

Crap In, Crap Out.

Well the same applies to natural language processing and sentiment analysis. Cleaning text is a little different to regular data cleaning in that, well, you’re dealing with strings of text rather than records of data. Regardless, cleaning the data is still important. The four most common steps that are performed are;

Lowercasing all words
Removing punctuation
Removing stopwords
Removing trivial words (excessively short and frequent words e.g. A, I, us)
Performing these steps ensures that the text is as relevant as possible to the message that the actual review is trying to represent.

Without further adieu, let’s get cleaning.

First up, lowercasing all the words. The stop word list that is used in a couple of steps has all of the words in lowercase, by lowercasing all of the text in the reviews it means that words which are capitalised won’t be missed

In [35]:
#Lowercasing all words
#A key part of text analysis is tokenization, this is where blocks of text are split into their individual words
#and pieces of punctuation. The punctuation itself does not add much meaning when searching for a topic or trying 
#to ascertain sentiment, hence its best to strip it out.
# Lower case all words
df['review_lower'] = df['Review'].apply(lambda x: " ".join(x.lower() for x in x.split()))

C:\Users\Mahwish\Anaconda2020\lib\site-packages\ipykernel_launcher.py:6: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  


In [36]:
#punctuation
import string
punc = string.punctuation # to see list of punctuation

df['review_nopunc'] = df['review_lower'].str.replace('punc', '')        


C:\Users\Mahwish\Anaconda2020\lib\site-packages\ipykernel_launcher.py:5: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """


In [37]:
# Import stopwords
from nltk.corpus import stopwords
stop_words = stopwords.words('english')

# Remove Stopwords
df['review_nopunc_nostop'] = df['review_nopunc'].apply(lambda x: " ".join(x for x in x.split() if x not in stop_words))

C:\Users\Mahwish\Anaconda2020\lib\site-packages\ipykernel_launcher.py:6: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  


In [38]:
# Return frequency of values
freq= pd.Series(" ".join(df['review_nopunc_nostop']).split()).value_counts()[:30]

In [39]:
other_stopwords = ['get', 'us', 'see', 'use', 'said', 'asked', 'day', 'go' \
  'even', 'ive', 'right', 'left', 'always', 'would', 'told', \
  'get', 'us', 'would', 'get', 'one', 'ive', 'go', 'even', \
  'also', 'ever', 'x', 'take', 'let' ]

In [40]:
df['review_nopunc_nostop_nocommon'] = df['review_nopunc_nostop'].apply(lambda x: "".join(" ".join(x for x in x.split() if x not in other_stopwords)))

C:\Users\Mahwish\Anaconda2020\lib\site-packages\ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """Entry point for launching an IPython kernel.


# This should leave you with a data frame with reviews at each of the different stages of cleaning. We’ll eventually drop these and replace them with our final sentiment metrics so don’t fret if it looks a little messy for now.

In [41]:
df.head()

Name RatingValue DatePublished  \
0     Playa Cabana     neutral    2018-09-07   
1       Fieramosca     neutral    2018-06-19   
2  El Nahual Tacos    negative    2019-09-10   
3             Miku    negative    2019-12-10   
4     Good Hombres    positive    2019-11-10   

                                              Review  length  word_count  \
0  My friend and I decided on Mexican food for di...    1153         209   
1  Had a good dinner here last week! The interior...    1939         370   
2  To be completely honest. I was expecting more....     393          69   
3  I wanted sooo badly to love Miku ...Because of...    1370         255   
4  This place serves some really great tacos! Ser...     241          45   

   char_count  avg_word  stopword_coun  \
0        1153  4.410377             92   
1        1939  4.204301            175   
2         393  4.628571             24   
3        1370  4.319066            118   
4         241  4.377778             21   

                                        review_lower  \
0  my friend and i decided on mexican food for di...   
1  had a good dinner here last week! the interior...   
2  to be completely honest. i was expecting more....   
3  i wanted sooo badly to love miku ...because of...   
4  this place serves some really great tacos! ser...   

                                       review_nopunc  \
0  my friend and i decided on mexican food for di...   
1  had a good dinner here last week! the interior...   
2  to be completely honest. i was expecting more....   
3  i wanted sooo badly to love miku ...because of...   
4  this place serves some really great tacos! ser...   

                                review_nopunc_nostop  \
0  friend decided mexican food dinner popped near...   
1  good dinner last week! interior traditional it...   
2  completely honest. expecting more. came lunch ...   
3  wanted sooo badly love miku ...because high ra...   
4  place serves really great tacos! service quick...   

                       review_nopunc_nostop_nocommon  
0  friend decided mexican food dinner popped near...  
1  good dinner last week! interior traditional it...  
2  completely honest. expecting more. came lunch ...  
3  wanted sooo badly love miku ...because high ra...  
4  place serves really great tacos! service quick...

# STEP 4 – LEMMATIZE THE REVIEWS

Lemmatization is the process of transforming your natural text from english to lemming language.

I kid, it’s the process of translating words back to their base form.

Lemmatization Example

am, are, is would be lemmatized to be
car, cars, car’s, cars’ would be lemmatized to car
This cuts out the number of words that are available for analysis by combining similar forms into one base form. One of other processes that is commonly used to cut down the the number of unique words in natural text processing is a process called stemming.

Stemming shortens the number of unique words by removing common endings.

Example:

Caresses is stemmed to caress
Ponies is stemmed to poni
Some words can stand alone without the extended ending however as shown with the word ponies above, this is not always the case. In this case we’ll use lemmatization to shorten down our word lsit.

The text blob module provides a simple method to lemmatize the reviews.

In [42]:
# Import textblob
from textblob import Word

# Lemmatize final review format
df['cleaned_review'] = df['review_nopunc_nostop_nocommon']\
.apply(lambda x: " ".join([Word(word).lemmatize() for word in x.split()]))

C:\Users\Mahwish\Anaconda2020\lib\site-packages\ipykernel_launcher.py:6: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  


#### There is an example of the change from base reviw to clean and lemmatized review

In [43]:
print('Base Review\n' ,df ["Review"][0])
print("\n------------------------------")
print("Cleaned and Lemmatized Review \n" ,df["cleaned_review"][0])

Base Review
 My friend and I decided on Mexican food for dinner and this popped up as nearby where we were at the time, Yorkville/Annex area. As we walked in, immediately we're greeted and seated in a tight space for 2. The menu was explained to us and water was given, then we were left for a few minutes to decide. 

We decided on salsa fresca & chips, chicken enchiladas suizas and short rib emmoladas along with a bottle each of corona and negra cervezas.

Salsa fresca and chips arrived first, the tomato salsa was delicious, the morita chile seemed to have been grilled before it was mashed together with tomatoes and red onions and went well with the crispy chips. (If you come for anything at all, this would be it and definitely skip the guac as it didn't seem appetizing.)

My friend and I shared the chicken enchiladas suizas and short rib emmoladas, while tasty, I could've gotten it somewhere else for cheaper. 

Because of how tight the space is, I was able to look around at what the o

# Sentiment Analysis

Finally, onto the sentiment analysis.

After all that cleaning actually performing the sentiment analysis is really quite straight forward. The textblob module again comes in quite handy for this task and returns not only sentiment metric but also a subjectivity metric as well.

The polarity metric refers to the degree to which the text analysed is positive or negative, between a range of -1 to 1. A score of 1 means highly positive whereas -1 is considered well and truly negative.

# 

In [44]:
# Calculate polarity
from textblob import TextBlob
df['polarity'] = df['cleaned_review'].apply(lambda x: TextBlob(x).sentiment[0])

C:\Users\Mahwish\Anaconda2020\lib\site-packages\ipykernel_launcher.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  This is separate from the ipykernel package so we can avoid doing imports until


In [45]:
df.head()

Name RatingValue DatePublished  \
0     Playa Cabana     neutral    2018-09-07   
1       Fieramosca     neutral    2018-06-19   
2  El Nahual Tacos    negative    2019-09-10   
3             Miku    negative    2019-12-10   
4     Good Hombres    positive    2019-11-10   

                                              Review  length  word_count  \
0  My friend and I decided on Mexican food for di...    1153         209   
1  Had a good dinner here last week! The interior...    1939         370   
2  To be completely honest. I was expecting more....     393          69   
3  I wanted sooo badly to love Miku ...Because of...    1370         255   
4  This place serves some really great tacos! Ser...     241          45   

   char_count  avg_word  stopword_coun  \
0        1153  4.410377             92   
1        1939  4.204301            175   
2         393  4.628571             24   
3        1370  4.319066            118   
4         241  4.377778             21   

                                        review_lower  \
0  my friend and i decided on mexican food for di...   
1  had a good dinner here last week! the interior...   
2  to be completely honest. i was expecting more....   
3  i wanted sooo badly to love miku ...because of...   
4  this place serves some really great tacos! ser...   

                                       review_nopunc  \
0  my friend and i decided on mexican food for di...   
1  had a good dinner here last week! the interior...   
2  to be completely honest. i was expecting more....   
3  i wanted sooo badly to love miku ...because of...   
4  this place serves some really great tacos! ser...   

                                review_nopunc_nostop  \
0  friend decided mexican food dinner popped near...   
1  good dinner last week! interior traditional it...   
2  completely honest. expecting more. came lunch ...   
3  wanted sooo badly love miku ...because high ra...   
4  place serves really great tacos! service quick...   

                       review_nopunc_nostop_nocommon  \
0  friend decided mexican food dinner popped near...   
1  good dinner last week! interior traditional it...   
2  completely honest. expecting more. came lunch ...   
3  wanted sooo badly love miku ...because high ra...   
4  place serves really great tacos! service quick...   

                                      cleaned_review  polarity  
0  friend decided mexican food dinner popped near...  0.066190  
1  good dinner last week! interior traditional it...  0.209410  
2  completely honest. expecting more. came lunch ...  0.348611  
3  wanted sooo badly love miku ...because high ra...  0.246273  
4  place serf really great tacos! service quick g...  0.626667

In [46]:
df_new = df[['Name','RatingValue','DatePublished','Review','cleaned_review']]
df_new

Name RatingValue DatePublished  \
0             Playa Cabana     neutral    2018-09-07   
1               Fieramosca     neutral    2018-06-19   
2          El Nahual Tacos    negative    2019-09-10   
3                     Miku    negative    2019-12-10   
4             Good Hombres    positive    2019-11-10   
..                     ...         ...           ...   
469                 Eataly     neutral    2020-03-07   
470                Terroni    negative    2019-10-20   
471  Hacienda las Américas    positive    2019-02-10   
472        Enoteca Sociale     neutral    2019-05-09   
473  Hong Kong Bistro Cafe    negative    2019-12-31   

                                                Review  \
0    My friend and I decided on Mexican food for di...   
1    Had a good dinner here last week! The interior...   
2    To be completely honest. I was expecting more....   
3    I wanted sooo badly to love Miku ...Because of...   
4    This place serves some really great tacos! Ser...   
..                                                 ...   
469  A group of us went for dinner at La Piazza wit...   
470  This restaurant is really for the experience. ...   
471  This place is a little gem. The food here is d...   
472  Enoteca is a nice spot with great drinks and g...   
473  Just prepared that you may wait more than 50 m...   

                                        cleaned_review  
0    friend decided mexican food dinner popped near...  
1    good dinner last week! interior traditional it...  
2    completely honest. expecting more. came lunch ...  
3    wanted sooo badly love miku ...because high ra...  
4    place serf really great tacos! service quick g...  
..                                                 ...  
469  group went dinner la piazza within eataly. not...  
470  restaurant really experience. lovely ambience ...  
471  place little gem. food delicious. say price qu...  
472  enoteca nice spot great drink great vibe grab ...  
473  prepared may wait 50 min food serving. came tw...  

[474 rows x 5 columns]

# Spliting for the dataset

In [54]:
from sklearn.model_selection import train_test_split
train_data, valid_data = train_test_split(df_new, test_size=0.2)

In [55]:

train_data.to_csv(r'C:\Users\Maha\training.csv')
valid_data.to_csv(r'C:\Users\Maha\valid.csv')

In [56]:
print(train_data.shape)
print(valid_data.shape)

(379, 5)
(95, 5)


In [57]:
X_train = train_data["cleaned_review"]
print(X_train.head())

X_valid = valid_data["cleaned_review"]
print(X_valid.head())

y_train = train_data["RatingValue"]
print(y_train)

y_valid = valid_data["RatingValue"]
print(y_valid)



466          came short male server buzz cut glass rude.
304    love ramen every winter i'm japan . dying try ...
261    i've mexico lindo bookmarked quite time now, s...
28     quetzal interesting. there's big huge space go...
146    really good tacos. must try bulgogi beef tat i...
Name: cleaned_review, dtype: object
343    love el trompo server nice, frozen margarita a...
462    tell truly nothing bad say place truly mean . ...
463    never fully understood taco genre life changed...
328    neighborhood decide eat others situation bust ...
179    absolutely terrible food worse attitude servic...
Name: cleaned_review, dtype: object
466    negative
304    negative
261    positive
28      neutral
146    positive
         ...   
112     neutral
171     neutral
299    positive
72      neutral
183    negative
Name: RatingValue, Length: 379, dtype: object
343    positive
462    positive
463    positive
328    negative
179    negative
         ...   
110    positive
403    positive
177    n

## CountVectorizer

In [58]:
from sklearn.feature_extraction.text import CountVectorizer
count_vect = CountVectorizer()
X_train_counts = count_vect.fit_transform(X_train)
X_train_counts.shape

(379, 4504)

### TFidf (Term Frequency inverse document freqency)

In [63]:
from sklearn.feature_extraction.text import TfidfTransformer
tf_transformer = TfidfTransformer(use_idf=True).fit(X_train_counts)
X_train_tf = tf_transformer.transform(X_train_counts)
X_train_tf.shape

(379, 4504)

In [64]:
tfidf_transformer = TfidfTransformer()
X_train_tfidf = tfidf_transformer.fit_transform(X_train_counts)
X_train_tfidf.shape

(379, 4504)

## Training the data using training data (Naive Bayes Classifier)

In [65]:
from sklearn.naive_bayes import MultinomialNB
clf = MultinomialNB().fit(X_train_tfidf,y_train )

In [66]:
#validating data
X_new_counts = count_vect.transform(X_valid)
X_new_tfidf = tfidf_transformer.transform(X_new_counts)

In [67]:
predicted = clf.predict(X_new_tfidf)

## Building a pipeline



In order to make the vectorizer => transformer => classifier easier to work with, scikit-learn provides a Pipeline class that behaves like a compound classifier:

In [68]:
from sklearn.pipeline import Pipeline
text_clf = Pipeline([
    ('vect', CountVectorizer()),
    ('tfidf', TfidfTransformer()),
    ('clf', MultinomialNB()),
])

In [69]:
text_clf.fit(X_train,y_train)

Pipeline(memory=None,
         steps=[('vect',
                 CountVectorizer(analyzer='word', binary=False,
                                 decode_error='strict',
                                 dtype=<class 'numpy.int64'>, encoding='utf-8',
                                 input='content', lowercase=True, max_df=1.0,
                                 max_features=None, min_df=1,
                                 ngram_range=(1, 1), preprocessor=None,
                                 stop_words=None, strip_accents=None,
                                 token_pattern='(?u)\\b\\w\\w+\\b',
                                 tokenizer=None, vocabulary=None)),
                ('tfidf',
                 TfidfTransformer(norm='l2', smooth_idf=True,
                                  sublinear_tf=False, use_idf=True)),
                ('clf',
                 MultinomialNB(alpha=1.0, class_prior=None, fit_prior=True))],
         verbose=False)

# Evaluating the predictive accuracy of the model on valid data

In [73]:
X_valid 
predicted = text_clf.predict(X_valid)
np.mean(predicted == y_valid )

0.7052631578947368

## Result:-

I have achieved 70.5% accuracy. Let’s see if I can do better with a linear support vector machine (SVM), which is widely regarded as one of the best text classification algorithms (although it’s also a bit slower than naïve Bayes). We can change the learner by simply plugging a different classifier object into our pipeline:

## SGD Classifier

In [74]:
from sklearn.linear_model import SGDClassifier
text_clf = Pipeline([
    ('vect', CountVectorizer()),
    ('tfidf', TfidfTransformer()),
    ('clf', SGDClassifier(loss='hinge', penalty='l2',
                          alpha=1e-3, random_state=42,
                          max_iter=5, tol=None)),
])

text_clf.fit(X_train,y_train)

predicted = text_clf.predict(X_valid)
np.mean(predicted == y_valid)

0.7473684210526316

## RESULT:-

I achieved 74.7% accuracy using the SVM. ```scikit-learn``` provides further utilities for more detailed performance analysis of the results:

In [76]:
List_Of_Categories_In_RatingValue=list(df_new['RatingValue'].value_counts().index)
List_Of_Categories_In_RatingValue

['negative', 'positive', 'neutral']

## Classification Reports

In [77]:
from sklearn import metrics
print(metrics.classification_report(y_valid, predicted,
    target_names=List_Of_Categories_In_RatingValue))

              precision    recall  f1-score   support

    negative       0.79      0.91      0.85        33
    positive       0.68      0.48      0.57        27
     neutral       0.74      0.80      0.77        35

    accuracy                           0.75        95
   macro avg       0.74      0.73      0.73        95
weighted avg       0.74      0.75      0.74        95



## Confusion Matrix results:

In [78]:
metrics.confusion_matrix(y_valid, predicted)

array([[30,  2,  1],
       [ 5, 13,  9],
       [ 3,  4, 28]], dtype=int64)